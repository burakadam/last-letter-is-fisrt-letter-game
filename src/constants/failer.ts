export const FAILER_PERCENTAGE = 30;
