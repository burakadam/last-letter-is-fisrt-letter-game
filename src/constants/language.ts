export const DEFAULT_LANGUAGE = 'tr-TR';
