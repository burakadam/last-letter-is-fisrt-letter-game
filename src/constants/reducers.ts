export const REDUCERS = {
  GAME: 'game',
};
