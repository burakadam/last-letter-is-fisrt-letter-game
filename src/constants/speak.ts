export const SPEAK_RATE = 0.75;
