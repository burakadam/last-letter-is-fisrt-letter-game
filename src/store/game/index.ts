export interface IErrors {
  listening: string | null;
  speaking: string | null;
}

export interface IGameState {
  gameStage: 'beggining' | 'game' | 'result';
  playerTurn: 'guest' | 'host' | null;
  isStarted: boolean;
  isSpeaking: boolean;
  isListening: boolean;
  hostAnswers: string[];
  activeHostAnswer: string;
  questAnswers: string[];
  activeGuestAnswer: string;
  answeredList: string[];
  resultText: string;
  errors: IErrors;
}

export * from './selectors';
export * from './slice';
export { default as gameReducer } from './slice';
