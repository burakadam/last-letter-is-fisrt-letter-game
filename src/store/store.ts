import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import { REDUCERS } from '@/constants/reducers';
import { gameReducer } from './game';

const rootReducer = combineReducers({
  [REDUCERS.GAME]: gameReducer,
});

export const store = configureStore({
  reducer: rootReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
