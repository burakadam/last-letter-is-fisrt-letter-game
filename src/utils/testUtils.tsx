import { PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function Wrapper({ children }: { children: React.ReactElement }) {
  return <>{children}</>;
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
