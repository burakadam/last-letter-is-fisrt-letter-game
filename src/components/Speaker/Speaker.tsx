import { useEffect } from 'react';

import { DEFAULT_LANGUAGE } from '@/constants/language';
import { SPEAK_RATE } from '@/constants/speak';
import { gameActions, gameSelectors } from '@/store/game';
import { useAppDispatch, useAppSelector } from '@/store/hooks';
import { _hostFindText } from '@/utils/hostFindText';

const synth = window.speechSynthesis || window.webkitSpeechRecognition;

const Speaker = () => {
  const dispatch = useAppDispatch();

  const playerTurn = useAppSelector(gameSelectors.playerTurn);
  const isSpeaking = useAppSelector(gameSelectors.isSpeaking);
  const activeHostAnswer = useAppSelector(gameSelectors.activeHostAnswer);
  const activeGuestAnswer = useAppSelector(gameSelectors.activeGuestAnswer);
  const answeredList = useAppSelector(gameSelectors.answeredList);

  useEffect(() => {
    if (playerTurn === 'host' && !isSpeaking && activeGuestAnswer !== null) {
      try {
        const answer = _hostFindText(answeredList, activeGuestAnswer);

        if (!answer) {
          dispatch(
            gameActions.setResultText({ text: 'Bilgisayar bir isim bulamadı.' })
          );
          dispatch(gameActions.resetToResult());
        } else {
          dispatch(gameActions.setGuestActiveAnswer({ answer: null }));

          const utterance = new SpeechSynthesisUtterance(answer);

          utterance.lang = DEFAULT_LANGUAGE;
          utterance.rate = SPEAK_RATE;

          synth.speak(utterance);

          utterance.onstart = () => {
            dispatch(gameActions.setHostActiveAnswer({ answer }));
            dispatch(gameActions.startSpeaking());
          };
          utterance.onend = () => {
            dispatch(gameActions.stopSpeaking());
            dispatch(gameActions.setPlayerTurn({ turn: 'guest' }));
            dispatch(gameActions.updateAnsweredList({ answer }));
            dispatch(gameActions.updateHostAnswers({ answer }));
          };
        }
      } catch (error) {
        dispatch(gameActions.stopSpeaking());
        dispatch(gameActions.errorSpeaking({ error: String(error) }));
      }
    }
  }, [activeGuestAnswer, dispatch, answeredList, isSpeaking, playerTurn]);

  return (
    <>
      <h1
        className={`mb-4 font-bold text-lg ${
          playerTurn === 'host' ? 'text-red-600' : ''
        }`}
      >
        Benim sıram
      </h1>
      <p className="uppercase bg-green-400 p-2 text-white font-bold text-center mb-4 w-auto">
        {activeHostAnswer ? activeHostAnswer : '....'}
      </p>
      <div className="w-[200px] h-1 bg-slate-500 my-4" />
    </>
  );
};

export { Speaker };
