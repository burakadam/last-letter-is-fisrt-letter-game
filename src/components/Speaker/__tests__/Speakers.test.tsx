/* eslint-disable testing-library/no-render-in-setup */
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';

import { REDUCERS } from '@/constants/reducers';
import { IGameState, initialState } from '@/store/game';
import { renderWithProviders } from '@/utils/testUtils';

import { Speaker } from '../Speaker';

const MockInitialState: IGameState = {
  ...initialState,
  gameStage: 'game',
  playerTurn: 'host',
  isStarted: true,
  isSpeaking: true,
  hostAnswers: [''],
  activeHostAnswer: 'Burak',
};

describe('add speaker that finds names from data/names.json with or without given name last letter', () => {
  let container: HTMLElement;
  beforeEach(() => {
    ({ container } = renderWithProviders(<Speaker />, {
      preloadedState: {
        [REDUCERS.GAME]: MockInitialState,
      },
    }));
  });

  it('render component correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('find the name comes from mock state', () => {
    expect(screen.getByText('Burak')).toBeInTheDocument();
  });
});
