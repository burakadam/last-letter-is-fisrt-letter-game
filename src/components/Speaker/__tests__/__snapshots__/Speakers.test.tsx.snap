// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add speaker that finds names from data/names.json with or without given name last letter render component correctly 1`] = `
<div>
  <h1
    class="mb-4 font-bold text-lg text-red-600"
  >
    Benim sÄ±ram
  </h1>
  <p
    class="uppercase bg-green-400 p-2 text-white font-bold text-center mb-4 w-auto"
  >
    Burak
  </p>
  <div
    class="w-[200px] h-1 bg-slate-500 my-4"
  />
</div>
`;
